File: src/Web.Client/src/types/property.ts
// Mirrors the backend Domain.Properties enums (serialized as numbers).
export enum ListingType {
  Rent = 0,
  Sale = 1,
}

export enum PropertyType {
  Backroom = 0,
  Cottage = 1,
  Room = 2,
  House = 3,
  Flat = 4,
}

export enum PropertyStatus {
  Listed = 0,
  Rented = 1,
  Sold = 2,
  Withdrawn = 3,
}

export const listingTypeLabels: Record<ListingType, string> = {
  [ListingType.Rent]: 'To Rent',
  [ListingType.Sale]: 'For Sale',
};

export const propertyTypeLabels: Record<PropertyType, string> = {
  [PropertyType.Backroom]: 'Backroom',
  [PropertyType.Cottage]: 'Cottage',
  [PropertyType.Room]: 'Room',
  [PropertyType.House]: 'House',
  [PropertyType.Flat]: 'Flat',
};

export const propertyStatusLabels: Record<PropertyStatus, string> = {
  [PropertyStatus.Listed]: 'Listed',
  [PropertyStatus.Rented]: 'Rented Out',
  [PropertyStatus.Sold]: 'Sold',
  [PropertyStatus.Withdrawn]: 'Withdrawn',
};

export interface PropertyImage {
  id: string;
  url: string;
}

export interface PropertySummary {
  id: string;
  title: string;
  listingType: ListingType;
  propertyType: PropertyType;
  price: number;
  township: string;
  city: string;
  province: string;
  bedrooms: number;
  bathrooms: number;
  hasElectricity: boolean;
  waterIncluded: boolean;
  hasOwnEntrance: boolean;
  hasParking: boolean;
  createdAt: string;
  imageUrls: string[];
}

export interface PropertyDetail extends Omit<PropertySummary, 'imageUrls'> {
  ownerId: string;
  description: string;
  street: string;
  postalCode: string;
  status: PropertyStatus;
  updatedAt: string | null;
  images: PropertyImage[];
}

export interface SearchFilters {
  township?: string;
  listingType?: ListingType;
  propertyType?: PropertyType;
  minPrice?: number;
  maxPrice?: number;
  minBedrooms?: number;
}

export interface PagedResult<T> {
  items: T[];
  page: number;
  pageSize: number;
  totalCount: number;
  hasNextPage: boolean;
}

export interface CreateListingRequest {
  ownerId: string;
  title: string;
  description: string;
  listingType: ListingType;
  propertyType: PropertyType;
  price: number;
  street: string;
  township: string;
  city: string;
  province: string;
  postalCode: string;
  bedrooms: number;
  bathrooms: number;
  hasElectricity: boolean;
  waterIncluded: boolean;
  hasOwnEntrance: boolean;
  hasParking: boolean;
}

export function formatPrice(price: number, listingType: ListingType): string {
  const amount = new Intl.NumberFormat('en-ZA', {
    style: 'currency',
    currency: 'ZAR',
    maximumFractionDigits: 0,
  }).format(price);

  return listingType === ListingType.Rent ? `${amount} / month` : amount;
}

File: src/Domain/Properties/Property.cs
using SharedKernel;

namespace Domain.Properties;

public sealed class Property : Entity
{
    public Guid Id { get; set; }
    public Guid OwnerId { get; set; }
    public string Title { get; set; }
    public string Description { get; set; }
    public ListingType ListingType { get; set; }
    public PropertyType PropertyType { get; set; }
    public decimal Price { get; set; }
    public Address Address { get; set; }
    public int Bedrooms { get; set; }
    public int Bathrooms { get; set; }
    public bool HasElectricity { get; set; }
    public bool WaterIncluded { get; set; }
    public bool HasOwnEntrance { get; set; }
    public bool HasParking { get; set; }
    public PropertyStatus Status { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }

    public static Property Create(
        Guid ownerId,
        string title,
        string description,
        ListingType listingType,
        PropertyType propertyType,
        decimal price,
        Address address,
        int bedrooms,
        int bathrooms,
        bool hasElectricity,
        bool waterIncluded,
        bool hasOwnEntrance,
        bool hasParking,
        DateTime utcNow)
    {
        var property = new Property
        {
            Id = Guid.NewGuid(),
            OwnerId = ownerId,
            Title = title,
            Description = description,
            ListingType = listingType,
            PropertyType = propertyType,
            Price = price,
            Address = address,
            Bedrooms = bedrooms,
            Bathrooms = bathrooms,
            HasElectricity = hasElectricity,
            WaterIncluded = waterIncluded,
            HasOwnEntrance = hasOwnEntrance,
            HasParking = hasParking,
            Status = PropertyStatus.Listed,
            CreatedAt = utcNow
        };

        property.Raise(new PropertyListedDomainEvent(property.Id));

        return property;
    }

    public Result MarkAsRented(DateTime utcNow)
    {
        if (ListingType != ListingType.Rent)
        {
            return Result.Failure(PropertyErrors.NotARentalListing(Id));
        }

        if (Status != PropertyStatus.Listed)
        {
            return Result.Failure(PropertyErrors.NotListed(Id));
        }

        Status = PropertyStatus.Rented;
        UpdatedAt = utcNow;

        Raise(new PropertyRentedDomainEvent(Id));

        return Result.Success();
    }

    public Result MarkAsSold(DateTime utcNow)
    {
        if (ListingType != ListingType.Sale)
        {
            return Result.Failure(PropertyErrors.NotASaleListing(Id));
        }

        if (Status != PropertyStatus.Listed)
        {
            return Result.Failure(PropertyErrors.NotListed(Id));
        }

        Status = PropertyStatus.Sold;
        UpdatedAt = utcNow;

        Raise(new PropertySoldDomainEvent(Id));

        return Result.Success();
    }

    public Result Withdraw(DateTime utcNow)
    {
        if (Status != PropertyStatus.Listed)
        {
            return Result.Failure(PropertyErrors.NotListed(Id));
        }

        Status = PropertyStatus.Withdrawn;
        UpdatedAt = utcNow;

        Raise(new PropertyWithdrawnDomainEvent(Id));

        return Result.Success();
    }

    public Result Relist(DateTime utcNow)
    {
        if (Status == PropertyStatus.Sold)
        {
            return Result.Failure(PropertyErrors.AlreadySold(Id));
        }

        if (Status == PropertyStatus.Listed)
        {
            return Result.Failure(PropertyErrors.AlreadyListed(Id));
        }

        Status = PropertyStatus.Listed;
        UpdatedAt = utcNow;

        Raise(new PropertyRelistedDomainEvent(Id));

        return Result.Success();
    }
}
